### DISCUSSION TITLE
Key codes and mapping

### DISCUSSION THREAD

**Discussion by pamoroso** [Interlisp/medley#1817] [General]
In the August 26, 2024 implementation meeting we discussed the issues with using Medley with non English keyboards. Since we mentioned some functions and tools to analyze key codes and mapping, to make sure I got them right here's a list:

* `SHOW-KEYS` in #1789
* `\KEYNAMES`
* `KEYDOWNP`

However, I may have missed what tool to use to actually remap the keyboard.

For reference here's the value of `\KEYNAMES` on Medley Online:

```
((5 %% FIVE)
 (4 $ FOUR)
 (6 ~ SIX)
 (e E)
 (7 & SEVEN)
 (d D)
 (u U)
 (v V)
 (0 %) ZERO)
 (k K)
 (- %)
 (p P)
 (/ ?)
 (\ %| FONT LOOKS)
 (LF SAME)
 (BS <-)
 (3 %# THREE)
 (2 @ TWO)
 (w W)
 (q Q)
 (s S)
 (a A)
 (9 %( NINE)
 (i I)
 (x X)
 (o O)
 (l L)
 (%, <)
 (%' %")
 (%] })
 (BLANK-MIDDLE OPEN DBK-HELP)
 (BLANK-TOP KEYBOARD DBK-META)
 (1 ! ONE)
 (ESC ESCAPE ->)
 (TAB =>)
 (f F)
 (CTRL PROP'S EDIT)
 (c C)
 (j J)
 (b B)
 (z Z)
 (LSHIFT)
 (%. >)
 (; %:)
 (CR <-%|)
 (_ ^)
 (DEL DELETE)
 (SKIP NEXT)
 (r R)
 (t T)
 (g G)
 (y Y)
 (h H)
 (8 * EIGHT)
 (n N)
 (m M)
 (LOCK)
 (SPACE)
 (%[ {)
 (= +)
 (RSHIFT)
 (BLANK-BOTTOM STOP)
 (MOVE)
 (UNDO)
 (UTIL0 SUN-KEYPAD=)
 (UTIL1 SUN-KEYPAD/)
 (UTIL2 SUPER/SUB)
 (UTIL3 CASE)
 (UTIL4 STRIKEOUT)
 (UTIL5 KEYPAD2)
 (UTIL6 KEYPAD3 PGDN)
 (UTIL7 SUN-LF)
 (PAD1 LEFTKEY CAPSLOCK KEYPAD+)
 (PAD2 LEFTMIDDLEKEY NUMLOCK KEYPAD-)
 (PAD3 MIDDLEKEY SCROLLLOCK KEYPAD*)
 (PAD4 RIGHTMIDDLEKEY BREAK KEYPAD/ SUN-PAUSE)
 (PAD5 RIGHTKEY DOIT PRTSC)
 (LEFT RED MOUSERED)
 (RIGHT BLUE MOUSEBLUE)
 (MIDDLE YELLOW MOUSEYELLOW)
 (MARGINS)
 (K41 KEYPAD7 HOME)
 (K42 KEYPAD8)
 (K43 KEYPAD9 PGUP)
 (K44 KEYPAD4)
 (K45 KEYPAD5)
 (K46 SUN-LEFT-SPACE)
 (K47 KEYPAD6)
 (K48 RIGHT-COMMAND SUN-RIGHT-SPACE)
 (COPY)
 (FIND)
 (AGAIN)
 (HELP)
 (DEF'N EXPAND)
 (K4E KEYPAD1 END)
 (ALWAYS-ON-1)
 (ALWAYS-ON-2)
 (CENTER)
 (K52 KEYPAD0 INS)
 (BOLD)
 (ITALICS)
 (UNDERLINE)
 (SUPERSCRIPT)
 (SUBSCRIPT)
 (LARGER SMALLER)
 (K59 KEYPAD%| KEYPAD.)
 (K5A KEYPAD\ KEYPAD, SUN-F10)
 (K5B SUN-F11)
 (K5C SUN-F12)
 (DEFAULTS SUN-PROP)
 (K5E SUN-PRTSC)
 (K5F SUN-OPEN))
```

**Comment by MattHeffron** [comment]
Sort of related info:
I used the X program `xev -event keyboard` in my Windows 11 WSL 2 Ubuntu Bash to get the information for _almost_ every key on my keyboard. (Some keys are intercepted by Windows and X never gets them, like `PrtScn` and the "Windows key".)
Here is a file with the `xev` output from pressing each key. (I removed the irrelevant info: x/y coordinates, timestamp...)
For most of the keys, the `keysym` or `XLookupString` value identifies the key pressed.
Some depend on the value of the `state` bits:
0x01   Shift key (either) is down.
0x02   Caps Lock is active.
0x04   Control key (either) is down.
0x08   Alt key (either) is down.
0x10   NumLock is active.
0x20, 0x40, 0x80  unused.

[xev keyboard.txt](https://github.com/user-attachments/files/16775915/xev.keyboard.txt)

**Comment by nbriggs** [comment]
I was looking at the USB HID Keyboard code page - which gives you the scan codes for any USB keyboard, except that it tells you things like:

> The usage type of all key codes is Selectors (Sel), except for the modifier keys Keyboard Left Control (0x224) to Keyboard
> Right GUI (0x231) which are Dynamic Flags (DV).
> Note: A general note on Usages and languages: Due to the variation of keyboards from language to language, it is not feasible to specify exact key mappings for every language. Where this list is not specific for a key function in a language, the closest equivalent key position should be used, so that a keyboard may be modified for a different language by simply printing different keycaps. One example is the Y key on a North American keyboard. In Germany this is typically Z. Rather than changing the keyboard firmware to put the Z Usage into that place in the descriptor list, the vendor should use the Y Usage on both the North American and German keyboards. This continues to be the existing practice in the industry, in order to minimize the number of changes to the electronics to accommodate other languages.
> 

Not exactly the most helpful!  I'd like to know the USB scan codes for Paolo's two keys to the right of the "L" keys.
